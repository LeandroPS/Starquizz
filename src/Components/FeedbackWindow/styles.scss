@import '../../styles/colors.scss';


.feedback-window{
    width: 100%;
    height: 100%;
    position: fixed;
    top: 100%;
    left: 0;

    background-color: #fff;

    text-align: center;

    transition: top 0.3s;

    &.show{
        top: 0;
        transition: top 0.3s;
    }

    div.points{
        padding: 20px;
        background-color: $light-gray;
        display: inline-block;
        font-size: 30px;
        border-radius: 5px;

        .title{
            display: block;
        }

        .points{
            display: block;
            font-weight: bold;
        }
    }

    form{
        margin: 20px 0px;

        input{
            display: block;
            margin: 10px 0px;
            width: 100%;
            padding: 10px;
            border-radius: 5px;
            border: 1px solid $gray;
            box-sizing: border-box;
        }

        span.message{
            display: block;
            margin-bottom: 10px;
            color: green;
        }
    }
}