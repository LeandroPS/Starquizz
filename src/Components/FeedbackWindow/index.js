import React , {Component} from 'react'
import PropTypes from 'prop-types';

import './styles.scss';
import '../../styles/common.scss';

class FeedbackWindow extends Component{
    constructor(){
        super();
        this.state = {
            user:{ 
                name: "",
                email: "",
                points: ""
            },
            showMessage: false
        }
    }

    componentDidMount(){
        this.setState({points: this.props.points});
    }

    saveUserOnLocalStorage = (user) =>{
        window.localStorage.setItem('user', JSON.stringify(user));
    }

    handleChange = (field, value) =>{
        let user = this.state.user;
        user[field] = value;
        this.setState({user});
    }

    onSubmitUserForm = (e) =>{
        e.preventDefault();
        this.saveUserOnLocalStorage(this.state.user);
        this.setState({showMessage: true});
    }

    render(){
        const {points, show} = this.props;
        const {showMessage} = this.state;

        return(
            <div className={`feedback-window ${show&&'show'}`}>
                <div className="centered">
                    <h1>Quizz finalizado</h1>
                    <div className="points">
                        <span  className="title">
                            você fez
                        </span>
                        <span  className="points">
                            {points} Pontos!
                        </span>
                    </div>
                    <form onSubmit={this.onSubmitUserForm}>
                        <span>Preencha este formulário pra salvar a sua pontuação:</span>
                        <input 
                            type="text" 
                            className="name"
                            onChange={(e)=>this.handleChange('name', e.target.value)} 
                            placeholder="Nome" 
                            required/>
                        <input 
                            type="email" 
                            className="email"
                            onChange={(e)=>this.handleChange('email', e.target.value)} 
                            placeholder="E-mail" 
                            required/>
                        {showMessage&&<span className="message">Você salvou suas informações com sucesso</span>}
                        <button className="button">Salvar</button>
                    </form>
                </div>
            </div>
        );
    }
}

FeedbackWindow.propTypes = {
    points: PropTypes.number.isRequired,
    show: PropTypes.bool.isRequired
};

export default FeedbackWindow;