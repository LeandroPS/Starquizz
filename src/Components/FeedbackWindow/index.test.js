import React from 'react';
import ReactDOM from 'react-dom';
import { create } from "react-test-renderer";
import { mount } from 'enzyme'; 
import FeedbackWindow from './';

describe('Feedback window', () => {
    const props = {
        points: 35,
        show: true
    }

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<FeedbackWindow {...props}/>, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders correctly', () => {
        const tree = create(<FeedbackWindow {...props}/>)
          .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('shows correct points when passed', () => {
        const component = mount(<FeedbackWindow {...props} />);
        expect(component.find('span.points').text()).toContain('35');
    });

    it('checks if both fields are filled', () => {
        const component = mount(<FeedbackWindow {...props} />);
        const form = component.find('form');

        const spy = jest.spyOn(component.instance(), 'onSubmitUserForm');

        form.simulate('submit');

        expect(spy).not.toHaveBeenCalled();

    });

    it('calls onSubmit when both fields are filled', () => {
        const component = mount(<FeedbackWindow {...props} />);
        const nameInput = component.find('input.name');
        const emailInput = component.find('input.email');
        const form = component.find('form');

        const spy = jest.spyOn(component.instance(), 'onSubmitUserForm');

        nameInput.simulate('change', { target: { value: 'Leandro' } });
        emailInput.simulate('change', { target: { value: 'leandro.pires.souza@gmail.com' } });
        form.simulate('submit');

        expect(spy).toHaveBeenCalled();
    });

    it('show message when saved', () => {
        const component = mount(<FeedbackWindow {...props} />);
        const nameInput = component.find('input.name');
        const emailInput = component.find('input.email');
        const form = component.find('form');

        nameInput.simulate('change', { target: { value: 'Leandro' } });
        emailInput.simulate('change', { target: { value: 'leandro.pires.souza@gmail.com' } });
        form.simulate('submit');

        expect(component.find('span.message').length).toEqual(1);;
    });

});
