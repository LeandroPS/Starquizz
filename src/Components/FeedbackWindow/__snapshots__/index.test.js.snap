// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback window renders correctly 1`] = `
<div
  className="feedback-window show"
>
  <div
    className="centered"
  >
    <h1>
      Quizz finalizado
    </h1>
    <div
      className="points"
    >
      <span
        className="title"
      >
        você fez
      </span>
      <span
        className="points"
      >
        35
         Pontos!
      </span>
    </div>
    <form
      onSubmit={[Function]}
    >
      <span>
        Preencha este formulário pra salvar a sua pontuação:
      </span>
      <input
        className="name"
        onChange={[Function]}
        placeholder="Nome"
        required={true}
        type="text"
      />
      <input
        className="email"
        onChange={[Function]}
        placeholder="E-mail"
        required={true}
        type="email"
      />
      <button
        className="button"
      >
        Salvar
      </button>
    </form>
  </div>
</div>
`;
